[Done] Implement spare gaussian process regression

fehiepsi opened the pull request:
@fritzo @ysaatchi This pull request is an attempt to implement sparse approximation versions of GPR (#702). The main reference for notations is [Damianou's thesis](http://etheses.whiterose.ac.uk/9968/1/Damianou_Thesis.pdf). The supported methods are DTC, FITC, and VFE.

- [x] Implement model and guide for DTC, FITC, VFE.
- [x] Implement inference for the above sparse methods.
- [x] Add support for Kernel to quickly calculate diagonal of covariance matrices.
- [x] Add options to predict full cov matrix or just diagonal form, with noise or not noise (y or f), jitter.
- [x] Move SpareMultivariateNormal to distribution section (make another pull request for this to easy to test).
- [x] Add tests to detect if there are errors in the implementation.
- [x] Fix bug in tutorial and ~add a section on sparse regression~ (move this part to another pull request).

Note that the implementations of Likelihood module and Mean function module will be in some future pull requests.

fritzo commented on pyro/contrib/gp/models/sgpr.py:
```diff
@@ -0,0 +1,182 @@
+from .util import _matrix_triangular_solve_compat
+
+
+class _SparseMultivariateNormal(dist.MultivariateNormal):
```
Feel free to move this into `pyro.distributions.sparse_mvn` or something. That would make it easier to add distribution tests for it.

fehiepsi replied:
Sure, after making things work, I will refactor it.

fehiepsi commented:
@fritzo @ysaatchi I have completed the implementation of sparse regression models. There are several tests for `forward` pass of these models. And I add a sparse section in the GP tutorial to illustrate the `inference` step. Would you mind helping me review this pull request?

p/s: I don't know if the way I unify all three models "DTC", "FITC", "VFE" is standard or not (I don't have much background, just learn about GP around two months ago, and this implementation comes from what I understand from various literature - especially [this note](https://github.com/GPflow/GPflow/blob/master/doc/source/notebooks/SGPR_notes.ipynb) - after writing down all computation details). As far as I can see, it is pretty fast to inference. The math in the implementation is quite clear to follow because I use the standard notions.

fehiepsi commented:
@fritzo Travis fails at running the tutorial due to not detecting my "pyro" kernel. I don't know how to resolve it. :(

neerajprad commented:
> @fritzo Travis fails at running the tutorial due to not detecting my "pyro" kernel. I don't know how to resolve it. :(

Could you try reverting the changes to the metadata, in particular removing the kernel name `pyro` from kernelspec? It's trying to activate the named ipython kernel `pyro`, but that is only on your local system. I think you can also just run `python -m nbstripout --keep-output tutorial/source/gp.ipynb` which will revert the non-essential changes to the notebook:

```
  "metadata": {
    "kernelspec": {		    "kernelspec": {
 -   "display_name": "Python 2",		 +   "display_name": "Python (pyro)",
     "language": "python",		     "language": "python",
 -   "name": "python2"		 +   "name": "pyro"
    },		    },
```

fehiepsi commented:
@neerajprad Thanks for the tip!!! I just modified it.

fehiepsi commented:
After https://github.com/pytorch/pytorch/pull/4923, we can use `mean`, `variance` in Distributions. I suggest we update Travis with a newer build of pytorch.

fritzo commented:
@fehiepsi we're having issues building a new PyTorch wheel for testing, but we'll update as soon as we can.

neerajprad commented:
> After pytorch/pytorch#4923, we can use mean, variance in Distributions. I suggest we update Travis with a newer build of pytorch.

The wheels are updated, and these methods should be available now. I restarted the tests that were failing earlier.

fritzo commented on pyro/contrib/gp/models/sgpr.py:
```diff
@@ -0,0 +1,172 @@
+        Kuf = kernel(Xu, self.X)
+        Luu = Kuu.potrf(upper=False)
+
+        # Ref: "A Unifying View of Sparse Approximate Gaussian Process Regression"
```
Thanks for pointing to literature #662

fritzo commented on tutorial/source/gp.ipynb:
```diff
@@ -433,44 +436,188 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsUAAAFpCAYAAAB0/VUQAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4xLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvAOZPmwAAIABJREFUeJztvX+IZUl253eiKt+Pek96qT9qsORRzfZY1XTJVi+q7KqctkStqU2lUSofvTb2FjYobfDCww8btNjQ3kZkL9uNGWevczB4oYzYXbBlsUuBVqzRYksjbi1GYElVLY9kzY4k5GWNhxWeWYytlQ22Zzf8R9W5fV68iLgRceP+eO9+P9B0ZeZ798a ...
```
Why is the envelope so narrow around the inducing points? I would expect a plot of the estimated total variance at each point x, whereas it looks like you're plotting only the uncertainty in the mean (not the variance due to measurement error). Do you think it would be clearer to plot total uncertainty rather than uncertainty in the mean?

fritzo replied:
(this comment refers to the final figure)

fehiepsi replied:
Nice catch! The current implementation predicts $f*$, not $y*$. Although I have put the option: `noiseless=False`, I forgot to include it in the implementation. Will fix this soon!

ysaatchi commented on pyro/contrib/gp/models/gpr.py:
```diff
@@ -5,81 +5,94 @@
+        if noiseless:
+            Kff = Kff + self.jitter.expand(self.num_data).diag()
+        else:
+            Kff = Kff + self.noise.expand(self.num_data).diag()
```
This is correct, however, it is safer to add jitter regardless, as sometimes the MAP solution may be attracted to short lengthscales and very low noise (aka Level-2 overfitting :))

fehiepsi replied:
I think that it is better to address this when we start to put constraints on parameters (lengthscale, variance, noise,...).

ysaatchi commented on pyro/contrib/gp/models/gpr.py:
```diff
@@ -5,81 +5,94 @@
-        # L = torch.potrf(K, upper=False)
-        # alpha = torch.potrs(L, self.y, upper=False)
+        # loc = Kfs.T @ inv(Kff) @ y
+        loc = Lffinv_Kfs.t().matmul(Lffinv_y)
```
wow, cool, this works now?

fehiepsi replied:
For `torch.trtrs` cuda and grad: not yet. We make a wrapper here: https://github.com/uber/pyro/blob/dev/pyro/distributions/util.py#L274

ysaatchi commented on pyro/contrib/gp/models/sgpr.py:
```diff
@@ -0,0 +1,172 @@
+    :param pyro.contrib.gp.kernels.Kernel kernel: A Pyro kernel object.
+    :param pyro.contrib.gp.InducingPoints Xu: An inducing-point module for spare approximation.
+    :param torch.Tensor noise: An optional noise tensor.
+    :param str approx: One of approximation methods: 'DTC', 'FITC', and 'VFE' (default).
```
may be nice for newcomers to expand these acronyms

fehiepsi replied:
I address this in the description of the model. :)

ysaatchi commented on tests/contrib/gp/test_models.py:
```diff
@@ -4,21 +4,72 @@
+
+    gpr = GPRegression(X, y, kernel, noise=noise)
+    sgpr = SparseGPRegression(X, y, kernel, Xu, noise=noise)
+    sgpr_ftic = SparseGPRegression(X, y, kernel, Xu, approx="FITC", noise=noise)
```
typo: call it sgpr_fitc

fehiepsi replied:
yup! thanks!

fehiepsi commented:
@ysaatchi @fritzo I addressed your reviews in the last commit. Especially, I fixed a mistake while introducing the `noiseless` parameter for prediction. Now, we get a reasonable plot for sparse version.

@ysaatchi I already made a preliminary version of the variational approach. Will make a pull request for it soon. It is much cleaner than the exact implementation here and is able to address general likelihood! :)
